symengine/parser/sbml/sbml_parser.cpp

```diff
@@ -0,0 +1,386 @@
+}
+
+static RCP<const Boolean> vec_lt(const vec_basic &x)
+{
+    set_boolean b;
+    for (std::size_t i = 0; i + 1 < x.size(); ++i) {
+        b.insert(Lt(x[i], x[i + 1]));
+    }
+    return logical_and(b);
+}
+
+RCP<const Basic> SbmlParser::modulo(const RCP<const Basic> &a,
+                                    const RCP<const Basic> &b)
+{
+    return piecewise({{sub(a, mul(b, ceiling(div(a, b)))),
+                       logical_xor({Lt(a, zero), Lt(b, zero)})},
+                      {sub(a, mul(b, floor(div(a, b)))), boolTrue}});
+}
+
+RCP<const Basic> SbmlParser::functionify(const std::string &name)
+{
+    const static std::map<const std::string, const RCP<const Basic>>
+        zero_arg_functions = {{"plus", integer(0)},
+                              {"times", integer(1)},
+                              {"and", boolTrue},
+                              {"or", boolFalse},
+                              {"xor", boolFalse}};
+    std::string lname = lowercase(name);
+    auto l = zero_arg_functions.find(lname);
+    if (l != zero_arg_functions.end()) {
+        return l->second;
+    }
+    throw ParseError("Parsing Unsuccessful: Function '" + name
+                     + "' has no arguments");
+}
+
+RCP<const Basic> SbmlParser::functionify(const std::string &name,
+                                         vec_basic &params)
+{
+    const static std::map<
+        const std::string,
+        const std::function<RCP<const Basic>(const RCP<const Basic> &,
+                                             const RCP<const Basic> &)>>
+        double_arg_functions = {
+            {"minus", sub}, {"divide", div}, {"pow", pow},
+            {"power", pow}, {"root", root},  {"log", sbml_log},
+        };
+    const static std::map<
+        const std::string,
+        const std::function<RCP<const Basic>(const RCP<const Basic> &)>>
+        single_arg_functions
+        = {{"sin", sin},         {"cos", cos},        {"tan", tan},
+           {"cot", cot},         {"csc", csc},        {"sec", sec},
+
+           {"asin", asin},       {"acos", acos},      {"atan", atan},
+           {"asec", asec},       {"acsc", acsc},      {"acot", acot},
+
+           {"arcsin", asin},     {"arccos", acos},    {"arctan", atan},
+           {"arcsec", asec},     {"arccsc", acsc},    {"arccot", acot},
+
+           {"sinh", sinh},       {"cosh", cosh},      {"tanh", tanh},
+           {"coth", coth},       {"sech", sech},      {"csch", csch},
+
+           {"asinh", asinh},     {"acosh", acosh},    {"atanh", atanh},
+           {"asech", asech},     {"acoth", acoth},    {"acsch", acsch},
+
+           {"arcsinh", asinh},   {"arccosh", acosh},  {"arctanh", atanh},
+           {"arcsech", asech},   {"arccoth", acoth},  {"arccsch", acsch},
+
+           {"sqr", sqr},         {"sqrt", sqrt},      {"root", sqrt},
+           {"abs", abs},         {"exp", exp},        {"ln", single_casted_log},
+           {"log", log10},       {"log10", log10},    {"ceil", ceiling},
```

Comment:
Should "log" always be log10? It seems as if this is optional in SBML.

Reply:
You're right, `log10` is the default but it can optionally instead be `ln` or `error`. There are also other options that are not yet supported here (see http://model.caltech.edu/software/libsbml/5.18.0/docs/formatted/cpp-api/_l3_parser_8cpp.html#a77c72a78afff728a6c569790338f2085), as well as user-defined functions and units.
I would address these in a future PR (I'll open an issue with the missing options/features)

Reply:
I've listed the missing features here: #1786

Reply:
Ok, great!